drivers/lsm6dsox/lsm6dsox.py

```diff
@@ -0,0 +1,228 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+
+Basic example usage:
+
+import time
+from lsm6dsox import LSM6DSOX
+
+from machine import Pin, I2C
+lsm = LSM6DSOX(I2C(0, scl=Pin(13), sda=Pin(12)))
+
+while (True):
+    print('Accelerometer: x:{:>8.3f} y:{:>8.3f} z:{:>8.3f}'.format(*lsm.read_accel()))
+    print('Gyroscope:     x:{:>8.3f} y:{:>8.3f} z:{:>8.3f}'.format(*lsm.read_gyro()))
+    print("")
+    time.sleep_ms(100)
+"""
+import array
+
+
+class LSM6DSOX:
+    CTRL3_C = const(0x12)
+    CTRL1_XL = const(0x10)
+    CTRL8_XL = const(0x17)
+    CTRL9_XL = const(0x18)
+
+    CTRL2_G = const(0x11)
+    CTRL7_G = const(0x16)
+
+    OUTX_L_G = const(0x22)
+    OUTX_L_XL = const(0x28)
+    MLC_STATUS = const(0x38)
+
+    DEFAULT_ADDR = const(0x6A)
+    WHO_AM_I_REG = const(0x0F)
+
+    FUNC_CFG_ACCESS = const(0x01)
+    FUNC_CFG_BANK_USER = const(0)
+    FUNC_CFG_BANK_HUB = const(1)
+    FUNC_CFG_BANK_EMBED = const(2)
+
+    MLC0_SRC = const(0x70)
+    MLC_INT1 = const(0x0D)
+    TAP_CFG0 = const(0x56)
+
+    EMB_FUNC_EN_A = const(0x04)
+    EMB_FUNC_EN_B = const(0x05)
+
+    def __init__(
+        self,
+        i2c,
+        address=DEFAULT_ADDR,
+        gyro_odr=104,
+        accel_odr=104,
+        gyro_scale=2000,
+        accel_scale=4,
+        ucf=None,
+    ):
+        """Initalizes Gyro and Accelerator.
+        accel_odr: (0, 1.6Hz, 3.33Hz, 6.66Hz, 12.5Hz, 26Hz, 52Hz, 104Hz, 208Hz, 416Hz, 888Hz)
+        gyro_odr:  (0, 1.6Hz, 3.33Hz, 6.66Hz, 12.5Hz, 26Hz, 52Hz, 104Hz, 208Hz, 416Hz, 888Hz)
+        gyro_scale:  (245dps, 500dps, 1000dps, 2000dps)
+        accel_scale: (+/-2g, +/-4g, +/-8g, +-16g)
+        ucf: MLC program to load.
+        """
+        self.i2c = i2c
+        self.address = address
+
+        # check the id of the Accelerometer/Gyro
+        if self.__read_reg(WHO_AM_I_REG) != 108:
+            raise OSError("No LSM6DS device was found at address 0x%x" % (self.address))
+
+        # allocate scratch buffer for efficient conversions and memread op's
+        self.scratch_int = array.array("h", [0, 0, 0])
+
+        SCALE_GYRO = {250: 0, 500: 1, 1000: 2, 2000: 3}
+        SCALE_ACCEL = {2: 0, 4: 2, 8: 3, 16: 1}
+        # XL_HM_MODE = 0 by default. G_HM_MODE = 0 by default.
+        ODR = {
+            0: 0x00,
+            1.6: 0x08,
+            3.33: 0x09,
+            6.66: 0x0A,
+            12.5: 0x01,
+            26: 0x02,
+            52: 0x03,
+            104: 0x04,
+            208: 0x05,
+            416: 0x06,
+            888: 0x07,
+        }
+
+        gyro_odr = round(gyro_odr, 2)
+        accel_odr = round(accel_odr, 2)
+        assert gyro_odr in ODR, "Invalid sampling rate: %d" % accel_odr
+        assert gyro_scale in SCALE_GYRO, "invalid gyro scaling: %d" % gyro_scale
+
+        assert accel_odr in ODR, "Invalid sampling rate: %d" % accel_odr
+        assert accel_scale in SCALE_ACCEL, "invalid accelerometer scaling: %d" % accel_scale
```

Comment:
When optimisations are enabled in mpy-cross (eg -O3) assertions like this are removed from the code.  In such a case there would not be any error checking.

If the intention is to keep these error checks even when opt is enabled, then they need to be changed to: `if not <cond>: raise ValueError(...)`.

Reply:
Good point, I missed this will fix.